book/src/dev/rfcs/drafts/xxxx-async-rust.md

```diff
@@ -0,0 +1,438 @@
+
+At a higher level, Zebra also uses [`tower::Service`s](https://docs.rs/tower/0.4.1/tower/trait.Service.html),
+[`tower::Buffer`s](https://docs.rs/tower/0.4.1/tower/buffer/struct.Buffer.html),
+and our own [`tower-batch`](https://github.com/ZcashFoundation/zebra/tree/main/tower-batch)
+implementation.
+
+# Motivation
+[motivation]: #motivation
+
+Like all concurrent codebases, Zebra needs to obey certain constraints to avoid
+hangs. Unfortunately, Rust's tooling in these areas is still developing. So
+Zebra developers need to manually check these constraints during design,
+development, reviews, and testing.
+
+# Definitions
+[definitions]: #definitions
+
+- `hang`: a Zebra component stops making progress.
+- `constraint`: a rule that Zebra must follow to prevent `hang`s.
+- `CORRECTNESS comment`: the documentation for a `constraint` in Zebra's code.
+- `task`: an async task can execute code independently of other tasks, using
+        cooperative multitasking.
+- `deadlock`: a `hang` that stops an async task executing code.
+        For example: a task is never woken up.
+- `starvation` or `livelock`: a `hang` that executes code, but doesn't do
+        anything useful. For example: a loop never terminates.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you are designing, developing, or testing concurrent Zebra code, follow the
+patterns in these examples to avoid hangs.
+
+If you are reviewing concurrent Zebra designs or code, make sure that:
+- it is clear how the design or code avoids hangs
+- the design or code follows the patterns in these examples (as much as possible)
+- the concurrency constraints and risks are documented
+
+The [Reference](#reference-level-explanation) section contains in-depth
+background information about Rust async concurrency in Zebra.
+
+Here are some examples of concurrent designs and documentation in Zebra:
+
+## Registering Wakeups Before Returning Poll::Pending
+[wakeups-poll-pending]: #wakeups-poll-pending
+
+Here's a wakeup correctness example from
+[unready_service.rs](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/unready_service.rs#L43):
+
+<!-- copied from commit de6d1c93f3e4f9f4fd849176bea6b39ffc5b260f on 2020-04-07 -->
+```rust
+// CORRECTNESS
+//
+// The current task must be scheduled for wakeup every time we return
+// `Poll::Pending`.
+//
+//`ready!` returns `Poll::Pending` when the service is unready, and
+// the inner `poll_ready` schedules this task for wakeup.
+//
+// `cancel.poll` also schedules this task for wakeup if it is canceled.
+let res = ready!(this
+    .service
+    .as_mut()
+    .expect("poll after ready")
+    .poll_ready(cx));
+```
+
+## Avoiding Deadlocks when Aquiring Buffer or Service Readiness
+[readiness-deadlock-avoidance]: #readiness-deadlock-avoidance
+
+Here's an example from [#1735](https://github.com/ZcashFoundation/zebra/pull/1735)
```

Comment:
Some of the examples link to the PR where they were done while others into the file in a branch. For consistency i think we can link to both places in each case.

Reply:
Now that those PRs are merged I can make this change.

Reply:
```suggestion
Here's an example from [zebra-consensus/src/chain.rs](https://github.com/ZcashFoundation/zebra/blob/3af57ece7ae5d43cfbcb6a9215433705aad70b80/zebra-consensus/src/chain.rs#L73)
in [#1735](https://github.com/ZcashFoundation/zebra/pull/1735):
```